builtin/providers/azure/config.go

```diff
@@ -0,0 +1,63 @@
+package azure
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/management"
+)
+
+// Config is a struct which holds all the required information to access
+// Azure services.
+type Config struct {
+	PublishSettingsFilePath string
+	SubscriptionID          string
+	ManagementCert          []byte
+	ManagementUrl           string
+}
+
+// AzureClient contains all the handles required for managing Azure services.
+type AzureClient struct {
+	// the client which holds the authentification information
+	managementClient management.Client
+
+	// unfortunately; because of how Azure's network API works; doing networking operations
+	// concurrently is very hazardous, and we need a mutex.
+	mutex *sync.Mutex
```

Comment:
:smiley:

I fully understand how the API works with regards to virtual networks, and indeed the API for virtual networks cannot be used concurrently in a safe way.

But did you actually check the Terraform internals with regards to plugin communication as I suggested? I don't think you did, so let me explain...

You are using a mutex to guard concurrent use of a `management.Client` within the plugin, to prevent problems with API endpoints that do not support being used concurrently. But there is no concurrency within the plugin itself. The concurrency is build into the core of Terraform and it will create a new instance of the plugin (most likely in it's own goroutine) for every call that can be made concurrently.

So in this case if you create a Terraform config with two virtual networks in there, Terraform core will actually call the plugin twice, resulting in two separate instances of the plugin that will then both have their own mutex.

Considering all this, you'll now probably understand that having the mutex here will not solve your problem as it will only guard an already synchronous instance of the plugin, from using the `management.Client` concurrently.

I just noticed you'd made a comment about this same topic on my WIP PR, so let me move this discussion over there so we discuss this only once...